### ISSUE TITLE
non-member function template in class template is mistaken for redefinition.

### ISSUE THREAD

**Issue by RedFog** [clangd/clangd#443]
Example:
```
template<bool A>
struct B{
    template<bool left, bool right>
    friend int operator-(B<left> _left, B<right> _right){ // < position 1
        return 0;
    }
};

static B<true> a;
static B<false> c; // < position 2
static int b = c - a;
```

it shows:

> _Redefinition of 'operator-'.
> In instantiation of template class 'B<false>' requested here. (in position 2)
> Previous definition is here. (in position 1)_

but it should work and factly works in **clang-1103.0.32.59**. the declaration of this function in each instance actually has only one entity.

in
**clangd 10.0.0.
VSCode 1.46.
MacOS 10.15.5**

by the way, clangd is always out of work in class template, such as:
```
template<typename T, size_t stride>
struct MeowVector{
    std::vector<T> data;
    typedef typename std::vector<T>::iterator iterator;
    typedef typename std::vector<T>::const_iterator const_iterator;
    void insert(const_iterator pos, std::vector<std::vector<T>> const& container) noexcept{
        size_t _size = container.size();
        data.reserve(data.size() + _size * stride);
        data.insert(pos, _size * stride, T());
        for (int i = 0; i != _size; ++i)
            for (int j = 0; j != container[i].size(); ++j)
                pos[i * stride + j] = container[i][j];
    }
};
```
the ```container[i].size()``` can not be deduced, even though it is independent of the template arguments.
it is disappointing if it can not show members and types in class templates, at least independent ones.

**Comment by HighCommander4** [comment]
> but it should work and factly works in **clang-1103.0.32.59**

Can you clarify where you got this clang version from? The recent clang versions that I'm aware of have version numbers like 8.0, 9.0, 10.0, 11.0, and all of them produce a redefinition error for this code, as does gcc.

> the `container[i].size()` can not be deduced, even though it is independent of the template arguments.

Unfortunately, the fact that C++ templates can be specialized means that pretty much any expression involving a dependent type is going to be dependent. In this case `container[i].size()` is dependent because (for example) `std::vector` could be specialized for some `T` and the specialization could have a completely different declaration of `operator[]` (which then could have a different return type, etc.).

Clangd does have logic to heuristically resolve some dependent expressions by assuming that the templates involved are not specialized, but handling this particular case heuristically is very difficult. (See [this bug](https://bugs.eclipse.org/bugs/show_bug.cgi?id=472818) for an extensive investigation into a very similar case. It's in the context of a different IDE, but the challenges are language-level and apply equally to clangd.)

**Comment by RedFog** [comment]
> clang --version
> Apple clang version 11.0.3 (clang-1103.0.32.59)

I tried Xcode and VSCode to run this code, with clang -std=c++17, and it worked correctly. and in fact, I had ever written some code like this when I used msvc (VS2015, maybe -std=c++14?) in Windows before, and it still worked. before I opened this issue, I have asked someone and search about it in standard, but no results. so if it should be a redefinition, could you tell me some information about it in cpp standard?

> In this case container[i].size() is dependent because (for example) std::vector could be specialized for some T and the specialization could have a completely different declaration of operator[] (which then could have a different return type, etc.).

you are right. I forget that it depends on ```T```, something may be different in some specialization, like ```vector<bool>```.

but in the other hand... I can imagine the difficulty of knowing the correct type and its members by pattern matching in lots of weird templates, and factly I always write these templates which may break the intellisense. but, even if ```container<T>``` in ```template<typename T>class Klass``` can not be deduced clearly, what else could it do?

To say the least, could I have any chance to hint to it that the ```vector<T>``` in some my class templates will not be any of specializations or exactly one? we use SFINAE and traits to do it in cpp code, and it seems to be weird for clangd, so I think comment hint is effective enough.

**Comment by HighCommander4** [comment]
> > clang --version
> > Apple clang version 11.0.3 (clang-1103.0.32.59)
> 
> I tried Xcode and VSCode to run this code, with clang -std=c++17, and it worked correctly. and in fact, I had ever written some code like this when I used msvc (VS2015, maybe -std=c++14?) in Windows before, and it still worked. before I opened this issue, I have asked someone and search about it in standard, but no results. so if it should be a redefinition, could you tell me some information about it in cpp standard?

Thanks. I don't have access to Xcode to try this out, but I know they apply some custom patches to clang. Perhaps they have one that affects this behaviour?

As far as why this code is invalid: the issue is that the friend function's declaration does not reference the enclosing class's template parameter `A` at all. As a result, the friend defined in `B<false>` and the friend defined in `B<true>` are the exact same function (same signature), and you get an error about trying to define that function twice.

The error is easy to fix, by moving the definition out of line:

```c++
template<bool A>
struct B{
    template<bool left, bool right>
    friend int operator-(B<left> _left, B<right> _right);
};

int operator-(B<left> _left, B<right> _right){
    return 0;
}

static B<true> a;
static B<false> c;
static int b = c - a;
```

This time, `B<false>` and `B<true>` both _declare_ the same function (which is fine), but there is only one definition.

**Comment by HighCommander4** [comment]
> > In this case container[i].size() is dependent because (for example) std::vector could be specialized for some T and the specialization could have a completely different declaration of operator[] (which then could have a different return type, etc.).
> 
> you are right. I forget that it depends on `T`, something may be different in some specialization, like `vector<bool>`.
> 
> but in the other hand... I can imagine the difficulty of knowing the correct type and its members by pattern matching in lots of weird templates, and factly I always write these templates which may break the intellisense. but, even if `container<T>` in `template<typename T>class Klass` can not be deduced clearly, what else could it do?
> 
> To say the least, could I have any chance to hint to it that the `vector<T>` in some my class templates will not be any of specializations or exactly one? we use SFINAE and traits to do it in cpp code, and it seems to be weird for clangd, so I think comment hint is effective enough.

The issue here is not about assuming that `vector<T>` isn't specialized; our heuristics already assume that without having to be told.

Rather, the issue is knowing that the return type of `vector<T>::operator[]` is `T` (or more precisely, `T&`).

Let's have a look at [libstdc++'s `vector` implementation](https://github.com/gcc-mirror/gcc/blob/e362a897655e3b92949b65a2b53e00fb3ab8ded0/libstdc%2B%2B-v3/include/bits/stl_vector.h), and see how `vector<T>::operator[]` is declared:

```c++
      reference
      operator[](size_type __n) _GLIBCXX_NOEXCEPT
```

As you can see, the declared return type is `vector<T>::reference`. Let's see how that's declared:

```c++
      typedef typename _Alloc_traits::reference		reference;
```

What is `__Alloc_traits`? It's:

```c++
      typedef __gnu_cxx::__alloc_traits<_Tp_alloc_type>	_Alloc_traits;
```

What is `_Tp_alloc_type`?

```c++
      typedef typename _Base::_Tp_alloc_type		_Tp_alloc_type;
```

What is `_Base`?

```c++
      typedef _Vector_base<_Tp, _Alloc>			_Base;
```

Ok, how does `_Vector_base` define `_Tp_alloc_type`?

```c++
      typedef typename __gnu_cxx::__alloc_traits<_Alloc>::template
	rebind<_Tp>::other _Tp_alloc_type;
```

I could go on, but I think you get the idea: the fact that `vector<T>::reference` is in fact `T&` is obscured by a generous amount of metaprogramming that's difficult to see through.

One way to think about this is that the type `vector<T>::reference` is the result of a compile-time program (metaprogram) whose input is the type `T`.

For a concrete type, there is no problem: you plug in e.g. `T = int`, run the metaprogram, and eventually get the result that `vector<int>::reference = int&`.

However, with a dependent type like `T`, you can't just "run the metaprogram" to come up with the answer that `vector<T>::reference = T&`, because metaprograms are not meant to be run with dependent types as inputs.

**Comment by RedFog** [comment]
> Thanks. I don't have access to Xcode to try this out, but I know they apply some custom patches to clang. Perhaps they have one that affects this behaviour?
> 
> As far as why this code is invalid: the issue is that the friend function's declaration does not reference the enclosing class's template parameter A at all. As a result, the friend defined in B<false> and the friend defined in B<true> are the exact same function (same signature), and you get an error about trying to define that function twice.

ok I will change it. I assumed the maybe duplicate definition will be converted to just one entity before, which may be what Xcode and VS do. I think it's easier to change it to asymmetric parameters, if I define it outside the class template, a lot of exhausting template arguments should be written.

> The issue here is not about assuming that ```vector<T>``` isn't specialized; our heuristics already assume that without having to be told.
> 
> Rather, the issue is knowing that the return type of ```vector<T>::operator[]``` is ```T``` (or more precisely, ```T&```).

I got it. but, hardly a method defined in ```vector<T>``` is not dependent on these weird type definitions, so I still think it's not ideal if no help of ```vector<T>::operator[]``` or even the whole ```vector<T>``` is given in a class template. 

honestly, it's always necessary to use some TMP ways in a simple class template, such as ```std::enable_if_t<XXX>``` or just ```typedef typename XXX<YYY>::ZZZ ZZZ```, but in most cases, they have similar behavior.

so why not assume ```typename T = int``` or ```bool flag = true``` to get more informations? if it does not pass, choose another arguments and try again, until no predictable arguments work.

and furthermore, give me a chance to write something like:
```
// hint with float, 3
template<typename T, size_t stride>
class XXX;

// hint with true
template<bool>
class YYY;
```
to avoid the clangd not working when in a simple class template. is there some ways like it?

**Comment by HighCommander4** [comment]
(My apologies, I accidentally edited your last comment when I meant to reply to it... I wish GitHub didn't make that so easy to do accidentally. Anyways, I've restored your comment the best I could, and my answer follows.)

---

> I got it. but, hardly a method defined in ```vector<T>``` is not dependent on these weird type definitions, so I still think it's not ideal if no help of ```vector<T>::operator[]``` or even the whole ```vector<T>``` is given in a class template. 

Note, we do offer completions for members of `vector<T>` (which are taken for the primary template definition of `vector`). It's only for dependent types that we can't map to a physical class definition (like `vector<T>::reference`) that it fails.

I agree that this is not ideal. Unfortunately, as I've explained, we are limited by the nature of C++ templates. Newer languages have learned from C++'s example and designed template-like features (like Rust's generics) that are much easier to write tooling for.

> so why not assume ```typename T = int``` or ```bool flag = true``` to get more informations? if it does not pass, choose another arguments and try again, until no predictable arguments work.

This is an interesting suggestion. I can think of two potential issues with it:

 * Not all templates can be successfully instantiated with all template parameter values, and compilers aren't necessarily designed to recover from arbitrary instantiation attempts. So, if we try to speculatively instantiate e.g. `A<T>` with `T = int`, and that causes a hard (non-SFINAE) error inside `A`, I don't think clang is designed to be able to recover from that and pretend the instantiation attempt didn't happen. (I could be mistaken about this; perhaps others who are more knowledgeable can chime in and confirm or refute this.)

 * Even if we overcome the first problem, we can't necessarily generalize the result of a concrete metaprogram result to a dependent type. So suppose we compute that for some template `A`, `A<int>::type = int`. Is that enough for us to conclude that `A<T>::type = T`? Not necessarily! For all we know, the definition of `type` may boil down to "if sizeof(T) < 8 then `int` otherwise `long`". And what if we find `A<int>::type = float`? That tells us approximately nothing about what `A<T>::type` is.

> and furthermore, give me a chance to write something like:
> ```
> // hint with float, 3
> template<typename T, size_t stride>
> class XXX;
> 
> // hint with true
> template<bool>
> class YYY;
> ```
> to avoid the clangd not working when in a simple class template. is there some ways like it?

There is nothing like this right now, but this is another interesting suggestion, and it would address the issues mentioned above with the previous suggestion.

I do wonder how likely is it that people are willing to add comments to their code that are incantations specific to clangd, but I guess there is some precedent here, e.g. with clang-tidy's `// NOLINT`.

**Comment by HighCommander4** [comment]
For completeness, I'll mention what I think is a more principled solution to the problem of code completion inside C++ templates: leveraging C++20 Concepts. If a template parameter is constrained by a concept, we have a lot more information about what sorts of values will be passed as an argument, and we can leverage that to provide better completion.

We've already done some work in this area in #261. With some further work, and a conceptified standard library implementation, I believe the `vector<T>::reference` issue can be solved too.

**Comment by HighCommander4** [comment]
(Updating title of issue to reflect the second topic (code completion inside templates), as the first topic (related to friend functions) has been addressed and there is nothing to fix there.)

**Comment by RedFog** [comment]
> Not all templates can be successfully instantiated with all template parameter values, and compilers aren't necessarily designed to recover from arbitrary instantiation attempts. So, if we try to speculatively instantiate e.g. ```A<T>``` with ```T = int```, and that causes a hard (non-SFINAE) error inside ```A```, I don't think clang is designed to be able to recover from that and pretend the instantiation attempt didn't happen. (I could be mistaken about this; perhaps others who are more knowledgeable can chime in and confirm or refute this.)
> 
> Even if we overcome the first problem, we can't necessarily generalize the result of a concrete metaprogram result to a dependent type. So suppose we compute that for some template ```A```, ```A<int>::type = int```. Is that enough for us to conclude that ```A<T>::type = T```? Not necessarily! For all we know, the definition of ```type``` may boil down to "if sizeof(T) < 8 then int otherwise long". And what if we find ```A<int>::type = float```? That tells us approximately nothing about what ```A<T>::type``` is.

yes, that is a problem, and I know we can not completely solve this problem right now. I'm sorry that I don't know how clangd works clearly, so it's just a suggestion trying to get more informations.

concepts in C++20 is a great work to look forward to, but it's far from us now and even when it releases, there will be many developers who may not use it, because of the cost of compatibility and so on. a solution is still needed before C++20.

for the second problem, the main idea for me is, to **try lots of patterns** and 
- **do an approximate decision** (for ```A<T>::reference```, ```A<int>::reference = int&``` and ```A<float>::reference = float&``` inaccurately prove ```A<T>::reference = T&```), or
- **just give a hint** like ```A<int>::reference = int&```, or even
- **give up the incorrect type deduction, but offer an approximate member deduction** (for ```vector<vector<T>> container```, treat ```container[i]``` as ```vector<int>```, and show the members of ```vector<int>```, because of the guess of ```vector<vector<int>>```).

and then we consider this comment hint:

> ```
> // hint with float, 3
> template<typename T, size_t stride>
> class XXX;
> 
> // hint with true
> template<bool>
> class YYY;
> ```

it gives a chance to tell clangd how to deduce the types and members and give user informations that they need. If I know my class template ```A<T>``` is a container, I wish clangd to treat it as a container, such as deduce with ```T = int```, which I can hint clangd in the comment. it's almost useful if clangd tells me what ```A<int>``` is when I hint it ```T = int```. it's my duty to ensure ```T = int``` works.

in the other hand, STL is almost defined by cpp standard. why not do more assumptions by standard? such as ```vector<T>::reference = T&, unless T = bool```. I think it will be a big improvement for intellisense, because most of time we use STL instead of other template libraries. in other words, it's better than nothing will be given even if we use STL.

finally, it's just a personal suggestion. the above-mentioned are just fit for me, maybe other programmers have their own needs. it doesn't mean I think it must be like this. I just want something more can be done to solve this problem. and at least, I don't mind writing lots of hints in comment to tell clangd what I want, after all, I use template.

**Comment by HighCommander4** [comment]
> Rather, the issue is knowing that the return type of `vector<T>::operator[]` is `T` (or more precisely, `T&`).
> 
> Let's have a look at [libstdc++'s `vector` implementation](https://github.com/gcc-mirror/gcc/blob/e362a897655e3b92949b65a2b53e00fb3ab8ded0/libstdc%2B%2B-v3/include/bits/stl_vector.h), and see how `vector<T>::operator[]` is declared:

Out of curiosity, I looked up [libcxx's `vector` implementation](https://github.com/llvm/llvm-project/blob/fc81f48fdeebd9e7c23900a410a39d3ced26019d/libcxx/include/vector#L482), and there is considerably less template metaprogramming there. I think we may be able to enhance our heuristics to the point where they can see through libcxx's `vector` implementation (but not `libstdc++`'s one).

**Comment by fgelm01** [comment]
Hello! I'm having trouble getting code assistance to work inside of template function bodies. This does not appear to be related to complexity of includes. Consider this code sample:

```cpp
template <typename A, typename B>
struct Pair {
  A first;
  B second;
};

template <typename C>
struct Vector {
  C& front();
  C const& front() const;
};

template <typename D, typename E>
void test(Vector<Pair<D, E>> const& vec) {
  vec.front().  // my cursor position is here
}
```

This is the entire file, there are no includes. I expect suggestions to include `first [D]` and `second [E]`.

This works just fine in Eclipse:

![image](https://github.com/clangd/clangd/assets/1566237/4f7e6a49-53e7-404c-ac0a-6c539834296f)

The exact same suggestion in Kate using its LSP client and clangd:

![image](https://github.com/clangd/clangd/assets/1566237/01e04317-9e7c-477b-9bd0-a0285a7e817f)

Notice how clangd thinks `front()` returns `C&`. It has not substituted `C` with `Pair<D, E>`.

The command used by kate to invoke `clangd` is 

```
/usr/bin/clangd-15 \
    -log=error \
    --background-index \
    --limit-results=500 \
    --completion-style=detailed \
    --completion-parse=always \
    --all-scopes-completion \
    --include-ineligible-results \
    --parse-forwarding-functions \
    --use-dirty-headers \
    --malloc-trim \
    -j=8 \
    --query-driver=/usr/bin/c++,/usr/bin/g++,/usr/bin/clang++*\
    --function-arg-placeholders
```

The output of `--version`:
```sh
$ /usr/bin/clangd-15 --version
Debian clangd version 15.0.7
Features: linux+grpc
Platform: x86_64-pc-linux-gnu
```

**Comment by HighCommander4** [comment]
@fgelm01 code completion for the case you mention is not currently implemented in clangd. We have #451 open for it.

> Notice how clangd thinks `front()` returns `C&`. It has not substituted `C` with `Pair<D, E>`.

Right, this substitution step is exactly what's not implemented. There is some discussion of it in these two comments:
 * https://github.com/clangd/clangd/issues/227#issuecomment-565871316
 * https://github.com/clangd/clangd/issues/227#issuecomment-565873123

The second comment suggests that this substitution may be more challenging to perform with clang AST, than it is with Eclipse CDT's AST. I think it's nonetheless worth trying, and I'd like to do so at some point.

If you're interested in trying to implement this improvement, let me know, I can try to provide guidance.

**Comment by fgelm01** [comment]
Hello @HighCommander4. I'd be willing to give it a try, and any guidance would be appreciated. Please be aware that I have not worked on clang, llvm, or clangd in the past.

What is a good starting point? So far I have read over [Design of clangd](https://clangd.llvm.org/design/) and am currently browsing `SemaCodeComplete.cpp`.

**Comment by HighCommander4** [comment]
I glossed over this when I wrote the previous comment, but #451 tracks the equivalent of this for features that use `HeuristicResolver` (mostly go-to-definition), and leveraging this in other features like code completion requires some additional steps.

So, a roadmap for getting this to work for code completion would involve the following steps:

 1. Implement the substitution logic in HeuristicResolver. This is tracked in #451, and I have an (old) WIP patch for it which can be used as a starting point.
 2. Upstream HeuristicResolver to the clang libraries, so code outside of clangd (like SemaCodeComplete.cpp) can use it. This is tracked in https://github.com/clangd/clangd/discussions/1662. A concrete first step discussed there is to write dedicated tests that exercise the HeuristicResolver interface directly.
 3. Actually use HeuristicResolver in SemaCodeComplete.cpp. (This should be straightforward with the previous two pieces in place.)

(1) and (2) can be worked on in parallel. If you're interested in helping out with either one, please feel free to post to the respective issue thread and I'm happy to provide further guidance.

**Comment by 16bit-ykiko** [comment]
I found that clion can provide completion in this case. It can also provide inlay hint for the variable.

![template](https://github.com/user-attachments/assets/6541ebd6-c9c3-4bd2-a0c9-f10453764e95)

Given that CLion can achieve this, I believe clangd can achieve it as well. After reading the previous discussions, I have some basic ideas. Based on the primary template assumption, we can always simplify a complex dependent name type into a type that does not contain dependent names. For example:

```cpp
template<typename T>
struct A {
    using type = T&;
};

template<typename U>
void foo(typename A<U>::type ref);
// typename A<U>::type => U&
```
Or a **direct** dependent name:

```cpp
template<typename T>
struct A {
    using type = typename T::type;
};

template<typename U>
void foo(typename A<U>::type ref);
// typename A<U>::type => typename U::type
```

In recent days, I have been validating this idea and have achieved some initial success. The code is as follows:

```cpp
class DependentShaper {
    clang::ASTContext& context;

public:
    DependentShaper(clang::ASTContext& context) : context(context) {}

    /// replace template arguments in a dependent type
    /// e.g.
    /// ```cpp
    /// template <typename T1, typename T2>
    /// struct A {
    ///     using reference = T1&;
    /// };
    ///
    /// template <typename U1, typename U2>
    /// struct B {
    ///    using type = A<U1, U2>::reference;
    /// };
    /// ```
    /// we want to simplify `A<U1, U2>::reference` to `U1&`, then we need to call
    /// `replace(<T1&>, <T1 = U1, T2 = U2>)` to get `U1&`.
    clang::QualType replace(clang::QualType input,
                            llvm::ArrayRef<clang::TemplateArgument> originalArguments) {
        assert(type->isDependentType() && "type is not dependent");

        // if the input is still a dependent name type, we need to simplify it recursively
        while(auto type = llvm::dyn_cast<clang::DependentNameType>(input)) {
            input = simplify(type);
        }

        if(auto type = llvm::dyn_cast<clang::LValueReferenceType>(input)) {
            auto pointee = type->getPointeeType();

            while(auto elaboratedType = llvm::dyn_cast<clang::ElaboratedType>(pointee)) {
                pointee = elaboratedType->desugar();
            }

            // get the position of the dependent type in the template parameter list
            // e.g `T1` in <typename T1, typename T2>, the index is 0
            if(auto paramType = llvm::dyn_cast<clang::TemplateTypeParmType>(pointee)) {
                auto index = paramType->getIndex();
                auto argument = originalArguments[index];

                // create a new type that fit the replacement
                return context.getLValueReferenceType(argument.getAsType());
            } else if(auto defType = llvm::dyn_cast<clang::TypedefType>(pointee)) {
                return replace(defType->getDecl()->getUnderlyingType(), originalArguments);
            } else {
                pointee->dump();
                std::terminate();
            }
        }
        // TODO: handle other kinds

        return input;
    }

    clang::NamedDecl* lookup(const clang::ClassTemplateDecl* classTemplateDecl,
                             const clang::IdentifierInfo* identifier) {
        clang::CXXRecordDecl* recordDecl = classTemplateDecl->getTemplatedDecl();
        auto result = recordDecl->lookup(identifier);
        return result.front();
    }

    /// for a complex dependent type: `X<...>::name::name2::...::nameN`, we can resolve it recursively.
    /// so we only need to handle the `X<...>::name`, whose prefix is a template specialization type.
    clang::QualType simplify(const clang::TemplateSpecializationType* templateType,
                             const clang::IdentifierInfo* identifier) {
        // X is a class template or a type alias template
        auto templateDecl = templateType->getTemplateName().getAsTemplateDecl();
        if(auto classTemplateDecl = llvm::dyn_cast<clang::ClassTemplateDecl>(templateDecl)) {
            // lookup the identifier in the record decl
            auto namedDecl = lookup(classTemplateDecl, identifier);

            if(auto decl = llvm::dyn_cast<clang::TypeAliasDecl>(namedDecl)) {
                return replace(decl->getUnderlyingType(), templateType->template_arguments());
            } else if(auto decl = llvm::dyn_cast<clang::TypedefDecl>(namedDecl)) {
                return replace(decl->getUnderlyingType(), templateType->template_arguments());
            } else {
                namedDecl->dump();
            }

        } else if(auto aliasTemplateDecl = llvm::dyn_cast<clang::TypeAliasTemplateDecl>(templateDecl)) {
            // TODO:
        } else {
            templateDecl->dump();
        }
    }

    const clang::QualType simplify(const clang::NestedNameSpecifier* specifier,
                                   const clang::IdentifierInfo* identifier) {
        auto kind = specifier->getKind();
        switch(specifier->getKind()) {
            case clang::NestedNameSpecifier::Identifier: {
                const auto prefix = simplify(specifier->getPrefix(), specifier->getAsIdentifier());

                if(auto type = llvm::dyn_cast<clang::TemplateSpecializationType>(prefix)) {
                    return simplify(type, identifier);
                } else {
                    prefix->dump();
                }

                break;
            }

            case clang::NestedNameSpecifier::TypeSpec: {
                auto node = specifier->getAsType();

                if(auto type = llvm::dyn_cast<clang::TemplateTypeParmType>(node)) {
                    // represent a direct dependent name, e.g. typename T::^ name
                    // and can not be further simplified
                    // node->dump();
                    type->dump();
                } else if(auto type = node->getAs<clang::TemplateSpecializationType>()) {
                    // represent a dependent name that is a template specialization
                    // e.g. typename vector<int>::^ name, and can be further simplified
                    return simplify(type, identifier);
                } else {
                    node->dump();
                }

                break;
            }

            case clang::NestedNameSpecifier::TypeSpecWithTemplate: {
                llvm::outs() << "unsupported kind: " << kind << "\n";
                break;
            }

            default: {
                llvm::outs() << "unsupported kind: " << kind << "\n";
            }
        }
    }

    const clang::QualType simplify(const clang::DependentNameType* type) {
        return simplify(type->getQualifier(), type->getIdentifier());
    }
};
```

It can now handle the following code:

```cpp
template <typename T>
struct A {
    using reference = T&;
};

template <typename U>
struct B {
    using reference = A<U>::reference;
};

template <typename Z>
struct C {
    using type = B<Z>::reference;
};
```

`B<Z>::reference` can be simplified to `Z&`. Of course, there are still many edge cases that haven't been considered, but I believe the current direction is correct.

Some questions:

- In the above code, the process of replacing template arguments involves manually unpacking the `ReferenceType` and replacing it according to the index of the `TemplateTypeParmType`. I wonder if clang provides some built-in functions to achieve this.
- For code completion, `CodeCompletionContext` provides a function `getBaseType()` to get the type of `member.*` in `member`, which seems to allow us to collect completion information without modifying clang upstream code, after simplifying the dependent name ourselves.

**Comment by HighCommander4** [comment]
Yes, this is definitely doable with some work. The recursive "simplification" logic in your code example is similar to the recursive logic in `HeuristicResolver` which I think we'd want to extend (currently it doesn't do anything like your `replace` step).

>    * In the above code, the process of replacing template arguments involves manually unpacking the `ReferenceType` and replacing it according to the index of the `TemplateTypeParmType`. I wonder if clang provides some built-in functions to achieve this.

I believe a general version of this replacement, which takes an arbitrary type that references template parameters, and replaces the template parameters with template arguments provided as input, no matter where in the type they are located, is done by [Sema::SubstType](https://searchfox.org/llvm/rev/f916cb6184fb04b27ae7b867edcfd162c3a49694/clang/include/clang/Sema/Sema.h#13264).

The high-level idea of calling into `Sema` to perform substitution is also what [my old WIP patch](https://reviews.llvm.org/D93522) was based on, though the specific Sema function called in that patch is different.

>    * For code completion, `CodeCompletionContext` provides a function `getBaseType()` to get the type of `member.*` in `member`, which seems to allow us to collect completion information without modifying clang upstream code, after simplifying the dependent name ourselves.

I don't think I'm following his part -- could you elaborate on how we would implement heuristics without modifying `SemaCodeComplete.cpp`?

**Comment by 16bit-ykiko** [comment]
Sorry, I guessed wrong.

I notice that there a function named [getBaseType](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Sema/CodeCompleteConsumer.h#L407) which can be used to get the type of `member` in `member.^`. So I think we can use it to get the `dependentNameType` and simplify it, then we can provide some completion result ourselves based on the simplified type. However, after a actual try, I find it return meaningless result in dependent case.

For example:

```cpp
#include <vector>

template<typename T>
void foo(){
    typename std::vector<std::vector<T>>::reference vec;
    vec.^
}
```
the result is

```text
BuiltinType 0x55f69c416df0 '<dependent type>' dependent
```

The result actually makes no sense. So we still need to modify the `SemaCodeComplete.cpp` to achieve our goal.

By the way, I am a college student currently on summer break, and I have about a month of free time to study this issue. I hope I can eventually solve this problem :).

**Comment by HighCommander4** [comment]
Note that there isn't any problem with modifying SemaCodeComplete.cpp; we've done it before to make code completion enhancement to clangd. (Such modifications also have the side effect of improving code completion for other clients of SemaCodeComplete such as libclang.)

The only catch is that currently, we don't have access to `HeuristicResolver` in SemaCodeComplete.cpp. I'm hoping to change that in the near future as discussed in https://github.com/clangd/clangd/discussions/1662.

**Comment by 16bit-ykiko** [comment]
Ok, after a quick view of `SemaCodeComplete.cpp`, I have found it's very easy to support `HeuristicResolver` for it. the main logic is located at [getAsRecordDecl](https://github.com/llvm/llvm-project/blob/f6d1d6fe7b7f04db7f0a445c8bc3e63fcdcd8e0a/clang/lib/Sema/SemaCodeComplete.cpp#L5223-L5243). Currently, it can only handle `TemplateSpecializationType` and can not handle `DependentNameType`.

So, I will focus on extending `HeuristicResolver` in next days.
